core/trino-main/src/main/java/io/trino/cache/CommonSubqueriesExtractor.java

```diff
@@ -0,0 +1,941 @@
+                        .collect(toImmutableList()),
+                Optional.of(groupByColumns));
+
+        return zip(subplans.stream(), childAdaptations.stream(),
+                (subplan, childAdaptation) -> {
+                    Map<CacheColumnId, Symbol> columnIdMapping = createSubplanColumnIdMapping(subplan, commonColumnIds, Optional.of(childAdaptation));
+                    PlanNode commonSubplan = createSubplanAggregation(
+                            childAdaptation.getCommonSubplan(),
+                            commonProjections,
+                            groupByColumns,
+                            columnIdMapping);
+                    List<Expression> adaptationConjuncts = createAdaptationConjuncts(subplan, TRUE, ImmutableSet.of(), Optional.of(childAdaptation));
+                    return new CommonPlanAdaptation(
+                            commonSubplan,
+                            planSignature,
+                            childAdaptation,
+                            createAdaptationPredicate(adaptationConjuncts, createSymbolMapper(columnIdMapping)),
+                            createAdaptationAssignments(commonSubplan, subplan, columnIdMapping),
+                            columnIdMapping,
+                            adaptationConjuncts);
+                }).collect(toImmutableList());
+    }
+
+    private Expression extractCommonPredicate(List<CanonicalSubplan> subplans)
+    {
+        // When two similar subqueries have different predicates, e.g: subquery1: col = 1, subquery2: col = 2
+        // then common subquery must have predicate "col = 1 OR col = 2". Narrowing adaptation predicate is then
+        // created for each subquery on top of common subquery.
+        return normalizeOrExpression(
+                extractCommonPredicates(
+                        or(subplans.stream()
+                                .map(subplan -> and(
+                                        subplan.getConjuncts()))
+                                .collect(toImmutableList()))));
+    }
+
+    private static Set<Expression> extractIntersectingConjuncts(List<CanonicalSubplan> subplans)
+    {
+        return subplans.stream()
+                .map(subplan -> (Set<Expression>) ImmutableSet.copyOf(subplan.getConjuncts()))
+                .reduce(Sets::intersection)
+                .map(ImmutableSet::copyOf)
+                .orElse(ImmutableSet.of());
+    }
+
+    private static Map<CacheColumnId, CacheExpression> extractCommonProjections(
+            List<CanonicalSubplan> subplans,
+            Expression commonPredicate,
+            Set<Expression> intersectingConjuncts,
+            Optional<Set<CacheColumnId>> pullupColumns,
+            Optional<List<CommonPlanAdaptation>> childAdaptations)
+    {
+        // Extract common projections. Common (cached) subquery must contain projections from all subqueries.
+        // Pruning adaptation projection is then created for each subquery on top of common subplan.
```

Comment:
So in case there are several subplans in which we read the same table with very different projections, after extracting common projections we'll read all the projections in all places.
That could be expensive in certain cases IMO.
In the load alternative, should we load only the necessary projections instead of pruning afterwards?

Reply:
I think this shoudn't be a problem, because all table scans should have same `requiredConjuncts`, so technically they should read same rows

Reply:
I meant, performance-wise (but I guess we can improve that later)

Reply:
Do you mean all projections are fetched from caching system even if they are not needed? We probably could ask `CacheManager` for subset of columns only required by given subquery and only compute "common projections" if they are absent. This wasn't an issue with in-memory cache as the cost was 0.

Reply:
Yes, that's what I meant. Right, we can do something like that in the future